Add BUILD_LAZY_CUDA_LINALG option

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72306

When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

malfet committed:
Add BUILD_LAZY_CUDA_LINALG option

When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33992795).

ngimel commented:
Windows build errors are real

IvanYashchuk commented:
So what happens if `BUILD_LAZY_CUDA_LINALG=OFF` and `torch.linalg.cholesky` is called on a CUDA tensor?

IvanYashchuk commented on aten/src/ATen/native/cuda/LinearAlgebraStubs.cpp:
```diff
@@ -0,0 +1,184 @@
+#include <ATen/native/BatchLinearAlgebra.h>
+#if AT_MAGMA_ENABLED()
+#include <ATen/cuda/detail/CUDAHooks.h>
+#include <iostream>
```
This include is probably not needed.

IvanYashchuk commented:
So what happens now when `BUILD_LAZY_CUDA_LINALG=OFF` and the user calls linear algebra functions? Normal behavior or an error? Should there be a CI job with `BUILD_LAZY_CUDA_LINALG=OFF`?

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33992795).

Lezcano commented on aten/src/ATen/native/cuda/LinearAlgebraStubs.cpp:
```diff
@@ -0,0 +1,184 @@
+  return lib;
+}
+
+Tensor& _linalg_inv_out_helper_cuda(Tensor &result, Tensor& infos_lu, Tensor& infos_getri) {
+    if (!_linalg_inv_out_helper_cuda_impl) {
+          _linalg_inv_out_helper_cuda_impl = reinterpret_cast<decltype(&_linalg_inv_out_helper_cuda)>(getTorchLinalgLibrary().sym("_ZN2at6native32_linalg_inv_out_helper_cuda_implERNS_6TensorES2_S2_"));
+          TORCH_CHECK(_linalg_inv_out_helper_cuda_impl != nullptr, "Can't find");
+        }
+    return _linalg_inv_out_helper_cuda_impl(result, infos_lu, infos_getri);
+}
```
Why is this one different to all the others?

malfet replied:
Sorry, that was an accidental leftover from the previous iteration. Rectified.

lezcano reviewed (commented):
As discussed with @IvanYashchuk , we both agree that it would be good to have a small write-up (perhaps at the top of `LinearAlgebraStubs.cpp` detailing what are the steps that we need to take to add a new function, and possible problems we may find while doing so.

The comment below shows that, at least by the looks of it, one may find some highly non-trivial cases, and it's not clear to us when may this happen.

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/cuda/LinearAlgebraStubs.cpp:
```diff
@@ -0,0 +1,188 @@
+}
+
+std::tuple<Tensor, Tensor> _solve_helper_cuda(const Tensor& self, const Tensor& A) {
+    static auto impl = getTorchLinalgFunc<decltype(&_solve_helper_cuda)>("_ZN2at6native18_solve_helper_cudaERKNS_6TensorES3_");
```
Would it be simpler to have a `Hooks` class similar to `CUDAHooks` or `VariableHooks`. Then this would just looks like:
```
getTorchLinalgLibrary()->_solve_helper_cuda(self, A);
```

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33992795).

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33992795).

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33992795).

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet committed:
Update on "Add BUILD_LAZY_CUDA_LINALG option"


When enable, it will generate `torch_cuda_linalg` library, which would depend on cusolve and magma and registers dynamic bindings to it from LinearAlgebraStubs

Differential Revision: [D33992795](https://our.internmc.facebook.com/intern/diff/D33992795)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33992795).

ngimel commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.h:
```diff
@@ -65,4 +65,18 @@ void lu_factor_looped_cusolver(const Tensor& self, const Tensor& pivots, const T
 
+#if defined(BUILD_LAZY_CUDA_LINALG)
+namespace cuda { namespace detail {
+struct LinalgDispatch {
```
nit: can you add a comment here that these are old-style dispatches and people shouldn't add to this list, and instead prefer implementing with stubs right away?

ngimel reviewed (approved):
This is great!

janeyx99 commented:
Sorry, reverting as this is the most likely culprit of causing timeouts in our OSS CI slow tests. See https://hud2.pytorch.org/hud/pytorch/pytorch/master/2
